crates/host/src/module.rs

```diff
@@ -165,12 +177,38 @@ impl SerializedModuleCache {
         wasm: &[u8],
     ) -> Result<Arc<ModuleWithStore>, wasmer::RuntimeError> {
         let store = Store::new((self.cranelift)());
-        let module = Module::from_binary(&store, wasm)
-            .map_err(|e| wasm_error!(WasmErrorInner::Compile(e.to_string())))?;
-        let serialized_module = module
-            .serialize()
-            .map_err(|e| wasm_error!(WasmErrorInner::Compile(e.to_string())))?;
-        self.put_item(key, Arc::new(serialized_module));
+
+        let maybe_module_path = self.module_path(key);
+        let (module, serialized_module) = match maybe_module_path.as_ref().map(|module_path| {
+            // We do this the long way to get `Bytes` instead of `Vec<u8>` so
+            // that the clone when we both deserialize and cache is cheap.
+            let mut file = File::open(module_path)
+                .map_err(|e| wasm_error!(WasmErrorInner::Compile(e.to_string())))?;
+            let mut bytes_mut = BytesMut::new().writer();
+            std::io::copy(&mut file, &mut bytes_mut)
+                .map_err(|e| wasm_error!(WasmErrorInner::Compile(e.to_string())))?;
+            Ok::<bytes::Bytes, wasmer::RuntimeError>(bytes_mut.into_inner().freeze())
+        }) {
+            Some(Ok(serialized_module)) => (
+                unsafe { Module::deserialize(&store, serialized_module.clone()) }
```

Comment:
Any way around the unsafe block here? Maybe a comment to explain why it's needed if not

Reply:
@ThetaSinner no because it loads data into executable code, and it is documented upstream